src/plugins/data/public/ui/query_editor/language_selector.tsx

```diff
@@ -54,14 +61,22 @@ export const QueryLanguageSelector = (props: QueryLanguageSelectorProps) => {
 
   const languageOptions: Array<{ label: string; value: string }> = [];
 
-  languageService.getLanguages().forEach((language) => {
-    if (
-      (language && props.appName && !language.editorSupportedAppNames?.includes(props.appName)) ||
-      languageService.getUserQueryLanguageBlocklist().includes(language?.id)
-    )
-      return;
-    languageOptions.unshift(mapExternalLanguageToOptions(language!));
-  });
+  languageService
+    .getLanguages()
+    .filter((language) => {
+      if (datasetSupportedLanguages) {
+        return datasetSupportedLanguages.includes(language.id);
+      }
+      return true;
+    })
+    .forEach((language) => {
+      if (
+        (language && props.appName && !language.editorSupportedAppNames?.includes(props.appName)) ||
+        languageService.getUserQueryLanguageBlocklist().includes(language?.id)
+      )
+        return;
```

Comment:
Could this be in `filter()`, since it's basically just filtering it again but in the `forEach()` loop? Then, we can pass the `mapExternalLanguageOptions()` function into a `.map()` instead of a for loop. Not a blocking comment, just something that might make it a bit easier more readable.

Reply:
combined forEach and filter